melodies_monet/util/metcalc.py

```diff
@@ -0,0 +1,240 @@
+variables including dewpoint and relative humidity are not universally provided by the obs. 
+For models, wind speed is usually not directly provided. 
+
+This script calculates all of them in one place. 
+
+Users can calculate variables directly from the observations by using the extra_calc dictionary. Potential 
+temperature is the only variable with this supported capability. 
+
+See: control_aircraft_looping_AEROMMA_UFSAQM.yaml
+
+Author: Liam Thompson
+"""
+
+# will need to make this an optional dependency if we proceed in using this. 
+import metpy
+
+# import specific metpy libraries needed for each calculation
+from metpy.calc import dewpoint_from_specific_humidity
+from metpy.calc import relative_humidity_from_specific_humidity
+from metpy.calc import potential_temperature
+from metpy.calc import wind_speed
+
+# if future iterations want to calc dewpoint/relh on the observations, this can be done with other metpy
+# libraries. 
+
+# addtl libraries to make the world go round
+from metpy.units import units
+import metpy.constants as mconst
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+# calc dewpoint 
+def dewpoint(obj, varmap = None, output_key = "dewpoint"):
+    # grab variable names from the yaml 
+    pressure_key = varmap['pressure'] if varmap and 'pressure' in varmap else 'surfpres_pa'
+    specific_hum_key = varmap['specific_hum'] if varmap and 'specific_hum' in varmap else 'specific_hum'
+
+    pressure = obj[pressure_key]
+    specific_hum = obj[specific_hum_key]
+
+    dpt = (metpy.calc.dewpoint_from_specific_humidity(
+        pressure * units.Pa,
+        specific_hum * units("kg/kg")
+    )).metpy.convert_units("K")
+
+    dpt_np = dpt.astype("float64").values
+
+    # DEBUG
+    # make sure this is generic enough
+    # print("pressure dims:", pressure.dims)
+    # print("specific_hum dims:", specific_hum.dims)
+    # print("dpt_np shape:", dpt_np.shape)
+
+    # fix is needed in order to work with vert plots. Check dimensions before proceeding. 
+    if hasattr(obj, "coords") and hasattr(obj, "dims"):
+        # Use pressure dims if number of dims matches dpt_np
+        if len(pressure.dims) == dpt_np.ndim:
+            dims_to_use = pressure.dims
+        # use 4d case for vert plots
+        elif len(specific_hum.dims) == dpt_np.ndim:
+            dims_to_use = specific_hum.dims
+        else:
+            raise ValueError(
+                f"No matching dims for output: pressure.dims={pressure.dims}, "
+                f"specific_hum.dims={specific_hum.dims}, dpt_np.shape={dpt_np.shape}"
+            )
+
+        obj[output_key] = (dims_to_use, dpt_np)
+        obj[output_key].attrs["units"] = "K"
+        
+        return obj
+
+        
+# calc relative humidity
+def relh(obj, varmap=None, output_key="rel_hum"):
+    # grab variable names from the yaml or fall back to defaults
+    pressure_key = varmap["pressure"] if varmap and "pressure" in varmap else "surfpres_pa"
```

Comment:
Same deal as the dewpoint calculation. You may want to specify the pressure variable in the varmap because surfpres_pa will not work in the vertical calculation.

Reply:
This may fix the bias for relative humidity aloft in your examples too.